src/IssueViz.UnitTests/Editor/SelectedIssueTagging/Adornment/IssueLocationAdornmentTagger_CacheTests.cs

```diff
@@ -0,0 +1,177 @@
+            testSubject.CachedAdornments.Should().BeEquivalentTo(actual1);
+
+            // Create another adornment
+            var actual2 = testSubject.CreateOrUpdate(newLocViz2);
+
+            actual2.LocationViz.Should().Be(newLocViz2);
+            actual1.Should().NotBe(actual2);
+            testSubject.CachedAdornments.Should().BeEquivalentTo(actual1, actual2);
+        }
+
+        [TestMethod]
+        public void CreateOrUpdate_Exists_ReturnsExistingAdornment()
+        {
+            var testSubject = new CachingAdornmentFactory(ValidTextView);
+            var locViz1 = CreateLocationViz();
+            var locViz2 = CreateLocationViz();
+
+            // Populate the cache
+            var adornment1 = testSubject.CreateOrUpdate(locViz1);
+            var adornment2 = testSubject.CreateOrUpdate(locViz2);
+            testSubject.CachedAdornments.Should().BeEquivalentTo(adornment1, adornment2);
+
+            // Act
+            testSubject.CreateOrUpdate(locViz1).Should().Be(adornment1);
+            testSubject.CreateOrUpdate(locViz2).Should().Be(adornment2);
+
+            testSubject.CachedAdornments.Should().BeEquivalentTo(adornment1, adornment2);
+        }
+
+        [TestMethod]
+        public void CreateOrUpdate_Exists_AdornmentIsUpdated()
+        {
+            var lineSource1 = CreateFormattedLineSource(12d);
+            var lineSource2 = CreateFormattedLineSource(18d);
+            var textView = CreateWpfTextView(ValidSnapshot, lineSource1);
+            var testSubject = new CachingAdornmentFactory(textView);
+
+            var locViz = CreateLocationViz();
+
+            var actual1 = testSubject.CreateOrUpdate(locViz);
+            actual1.FontSize.Should().Be(12d); // sanity check
+
+            ChangedMockedLineSource(textView, lineSource2);
+
+            // Act
+            var actual2 = testSubject.CreateOrUpdate(locViz);
+            actual2.FontSize.Should().Be(18d);
+        }
+
+        [TestMethod]
+        public void RemoveUnused_EmptyCache_NoError()
+        {
+            var testSubject = new CachingAdornmentFactory(ValidTextView);
+
+            Action act = () => testSubject.RemoveUnused(new[] { ValidLocViz });
+            act.Should().NotThrow();
+        }
+
+        [TestMethod]
+        public void RemoveUnused_NoCurrentLocations_AllRemoved()
+        {
+            var testSubject = new CachingAdornmentFactory(ValidTextView);
+            testSubject.CreateOrUpdate(CreateLocationViz());
+            testSubject.CreateOrUpdate(CreateLocationViz());
+            testSubject.CreateOrUpdate(CreateLocationViz());
+            testSubject.CachedAdornments.Count.Should().Be(3); // sanity check
+
+            testSubject.RemoveUnused(Array.Empty<IAnalysisIssueLocationVisualization>());
+            testSubject.CachedAdornments.Should().BeEmpty();
+        }
+
+        [TestMethod]
+        public void RemoveUnused_AllLocationsAreCurrent_NoAdormentsRemoved()
```

Comment:
Maybe
`Refresh_LocationsAreUnchanged_NoAdormentsRemoved` ?
(I think it should be Refresh)

Reply:
I think the important point is whether the cached locations are current or not, not whether the supplied set of locations has changed or not. For me
`Refresh_AllCachedLocationsAreCurrent_NoAdormentsRemoved` is clearer.